Reworked beat_active control:

JoergAtGithub opened the pull request:
-Made CO beat_active read-only
-Set CO only if there's a state change
-Update from engine::updateIndicators instead of process, which isn't updated enough for a reliable beat indicator
-Added NoTolerance mode in beatgrid.cpp
-Don't calculate if deck is not playing or vinyl/jog is hold standstill (inside +-2.5ms tracktime at normal speed)
-Set beat_active at the beat not +-50ms before/after
-Reverse state
-Period of beat_active on is no longer hard coded 200ms, it's no 20% of the peat period

![grafik](https://user-images.githubusercontent.com/64457745/110168573-adba6480-7df7-11eb-8159-c84f52d58aab.png)

See also discussion at: https://mixxx.zulipchat.com/#narrow/stream/109171-development/topic/beat_active

JoergAtGithub committed:
Reworked beat_active control:
-Made CO beat_active read-only
-Update from engine::updateIndicators instead of process, which isn't updated enough for a reliable beat indicator
-Fixed accuracy issue in beatgrid.cpp (compare to 1/100 of a beat period can be severalsamples off)
-Don't calculate if deck is not playing
-Set beat_active at the beat not +-50ms before/after
-More states
-Period of beat_active on is no longer hard coded 200ms, it's no 20% of the peat period

Holzhaus commented:
Thanks. Please open a manual PR to document the new behavior. That will make review easier.

JoergAtGithub commented:
> 
> 
> Thanks. Please open a manual PR to document the new behavior. That will make review easier.

Documentation is in https://github.com/mixxxdj/manual/pull/347

JoergAtGithub committed:
Added switch, to select if findPrevNextBeats should compare with or without tolerance epsilon

JoergAtGithub committed:
Added testcases for NoTolerance mode of findPrevNextBeats

JoergAtGithub committed:
Merge remote-tracking branch 'upstream/main' into improve_beat_indicator

JoergAtGithub committed:
Improved if cases for NoTolerance code in findPrevNextBeats

daschuer commented:
I am unsure if this is now just right. What is your use case for -1 -0.5 and 0.5? 

The simple use case is just a beat LED. It can be controlled like 
if (beat_active) {
    set_led() 
} else {
    reset_led()
}

This works perfectly if we play forward and backward. But IMHO not too nice when scratching, because it will not reset after changing direction. My expectation is that I can cue by eye just like I can do by ear. Therefore a change at the beat is required. 

if (beat_active >= 1 || beat_active <= -1) {
    set_led() 
} else {
    reset_led()
}

will be a solution.

daschuer commented on src/engine/controls/clockcontrol.cpp:
```diff
@@ -30,27 +33,94 @@ void ClockControl::trackLoaded(TrackPointer pNewTrack) {
+
+            m_blinkIntervalSamples = (m_NextBeatSamples - m_PrevBeatSamples) * kBlinkInterval;
+        }
+        //qDebug() << "dRate:" << dRate << " m_lastPlayDirection:" << m_lastPlayDirection << " m_pCOBeatActive->get(): " << m_pCOBeatActive->get() << " currentSample: " << currentSample << " m_lastEvaluatedSample: " << m_lastEvaluatedSample << " m_PrevBeatSamples: " << m_PrevBeatSamples << " m_NextBeatSamples: " << m_NextBeatSamples << " m_blinkIntervalSamples: " << m_blinkIntervalSamples;
```
Please break these long lines.

JoergAtGithub replied:
> I am unsure if this is now just right. What is your use case for -1 -0.5 and 0.5?

The intention was to give the mapping developer all possibilities:
1)Search the beat by LED (therfore the LED must go of if direction changes)
2)Only show LED in normal playing direction.
3)LED On if position is close to the beat independent of direction

But maybe it's too much. If I remove the +- 0.5 states an use the value 0 instead, a simple
if (beat_active !== 0)
   set_led()
} else {
   reset_led()
}
would work for case 1.
Only case 3) would be impossible. The reason for case 3 is taht it's similar to the current behavio,

JoergAtGithub replied:
I will remove this line later - it's just for my own debugging.

daschuer reviewed (changes requested):
Thank you very much for taking care. 
The code looks good. Just a minor code style comment. 
do you the precommit hook?
https://github.com/mixxxdj/mixxx/wiki/Using-Git#set-up-automatic-code-checking

JoergAtGithub commented:
> I am unsure if this is now just right. What is your use case for -1 -0.5 and 0.5?

The intention was to give the mapping developer all possibilities:
1)Search the beat by LED (therfore the LED must go of if direction changes)
2)Only show LED in normal playing direction.
3)LED On if position is close to the beat independent of direction

But maybe it's too much. If I remove the +- 0.5 states an use the value 0 instead, a simple
if (beat_active !== 0)
set_led()
} else {
reset_led()
}
would work for case 1.
Only case 3) would be impossible. The reason for case 3 is taht it's similar to the current behavior

JoergAtGithub commented:
> do you the precommit hook?
No, I don't use it. But the pre-commit check in the GitHub actions passed. Isn't this the same check?

JoergAtGithub commented:
Regarding the state of this code:

- Slow searching of a position with vinyl and LED works very well now!
- But while playing, the indication is not always at the expected moment. At least no beat seems to be missing anymore.
   My laptop is not very powerfull, maybe this is the reason

daschuer commented:
I won't use 3) do you? 
If not, let's remove it to simplify the other two.

daschuer commented:
>> do you the precommit hook?

> No, I don't use it. But the pre-commit check in the GitHub actions passed. Isn't this the same check?

Yes, that should be the same. I have no Idea why it is not complaining about the long comments. 
My point is just to remove the annoyance of vertical scrolling during code reviews.

daschuer commented:
Did you test using a controller LED or a GUI LED? I can imagine that there is a significant jitter due to the processing of the midi messages. Do you have similar issues with the VU meter?

JoergAtGithub committed:
Merge remote-tracking branch 'upstream/main' into improve_beat_indicator

JoergAtGithub committed:
Merge remote-tracking branch 'upstream/main' into improve_beat_indicator

JoergAtGithub committed:
Moved the 5 states into an internal state machine
The CO returns now 0 for the cases of play direction changes while beat indication is on

JoergAtGithub commented:
> 
> 
> I won't use 3) do you?
> If not, let's remove it to simplify the other two.

I changed this, the states +-0.5 are no longer send to the CO.

JoergAtGithub committed:
Added the loop cases:
-quantized loop on beat
-loop smaller than beat distance
-loop without a beat inside

daschuer commented:
Since the old behavior is a bug for me, I would like to see this in 2.3. 
It his OK for anyone?

daschuer commented:
In this case, can you rebate it to 2.3? 
git rebase --onto upstream/2.3 4719d5e9529e39b64e46e3cf46fe06a6106e0e66~1

daschuer commented on src/engine/controls/bpmcontrol.cpp:
```diff
@@ -539,7 +539,7 @@ bool BpmControl::getBeatContext(const mixxx::BeatsPointer& pBeats,
     double dPrevBeat;
     double dNextBeat;
-    if (!pBeats->findPrevNextBeats(dPosition, &dPrevBeat, &dNextBeat)) {
+    if (!pBeats->findPrevNextBeats(dPosition, &dPrevBeat, &dNextBeat, false)) {
```
We can use her true and remove undoing the fix in getBeatContextNoLookup(). 
This is actually a bug in case of a beat map where the beats can be differently.

JoergAtGithub replied:
I changed this, but I don't know how to verify this.

daschuer commented on src/engine/controls/clockcontrol.cpp:
```diff
@@ -30,27 +37,150 @@ void ClockControl::trackLoaded(TrackPointer pNewTrack) {
 
     // TODO(XXX) should this be customizable, or latency dependent?
-    const double blinkSeconds = 0.100;
+    const double kBlinkInterval = 0.20; // LED is on 20% of the beat period
```
this both can become a constexpr

JoergAtGithub replied:
Done!

daschuer commented on src/engine/controls/clockcontrol.cpp:
```diff
@@ -30,27 +37,150 @@ void ClockControl::trackLoaded(TrackPointer pNewTrack) {
+        const double loop_end_position = m_pLoopEndPosition->get();
+
+        if ((m_PrevBeatSamples < loop_start_position) && (m_NextBeatSamples >= loop_end_position)) {
+            // No beat inside loop -> show beat indication at loop_start_position
```
nice idea

daschuer reviewed (changes requested):
Works already good. I have left some commit. For the test, I have mapped the kill buttons, and it is really fun. 

Can we switch of the indicator if the track is paused on a cue point? Currently is is sometimes on and sometimes off when at a cue point, which is placed on a beat. 

Unfortunately the backward direction does not work instantly. Because in Mixxx we have the definition that ON = x > 0. A rocker switch would have the values 0 1 2 .. 
Would it be OK to use 2 for backwards, than it is full compatible to a three state GUI button.  

Especially with long audio buffers, the jitter is visible. This can be fixed by moving the ClockControl::updateIndicators into the waveform update loop which is synced with the display refresh rate. Via the visual play position you have access to the sample that is currently played, and not the one that have been processed as in the current solution. 

Optimizing this for the GUI did not help much with the latency that happens until the signal is at the controller though. So the full blown solution has to consider both paths individual.  

This is obviously no longer a  bug fix and a mayor refactoring. Maybe you should just add a note about it in ClockControl::updateIndicators for the case we want to pick it up in future.

JoergAtGithub committed:
Removed getBeatContextNoLookup() by using findPrevNextBeats without tolerance/snap behavior

JoergAtGithub committed:
Renamed NoTolerance to snapToNearBeats and inverted the bool, to match the new term

JoergAtGithub commented on src/track/beatgrid.h:
```diff
@@ -52,8 +52,10 @@ class BeatGrid final : public Beats {
+
+            double* dpPrevBeatSamples,
+            double* dpNextBeatSamples,
+            bool NoTolerance) const override;
```
Done!

JoergAtGithub committed:
const -> constexpr

JoergAtGithub committed:
Added ToDo comment about moving the updateIndicators to waveform update loop.

daschuer commented on src/engine/controls/clockcontrol.cpp:
```diff
@@ -30,27 +43,145 @@ void ClockControl::trackLoaded(TrackPointer pNewTrack) {
+    /* This method sets the control beat_active is set to the following values:
+    * +1.0 --> Forward playing, set at the beat and set back to 0.0 at 20% of beat distance
+    *  0.0 --> No beat indication (ouside 20% area or play direction changed while indication was on)
+    * -1.0 --> Reverse playing, set at the beat and set back to 0.0 at -20% of beat distance
```
Did you consider to change this to two? This way we can directly bind it to a three state button with "<" and ">" or similar icons.

JoergAtGithub replied:
Done!

daschuer commented on src/engine/controls/clockcontrol.cpp:
```diff
@@ -30,27 +43,145 @@ void ClockControl::trackLoaded(TrackPointer pNewTrack) {
+            }
+        }
+        m_lastPlayDirection = false; // Reverse
     }
```
I think I would prefer  
```suggestion
    } else {
        m_pCOBeatActive->forceSet(0.0);
    }
```
Because I experience that the LED is sometimes on and sometimes off when paused at a cue point.

JoergAtGithub replied:
I will look how to solve the cue point issue. The proposed else branch will not do this. It will never be reached, because this is captured before, to not switch the indicator off due the jitter of the DVS pitch.

JoergAtGithub replied:
I applied a fix for this, which should not break the use case of searching the beat position using DVS or jog. Please test, if this solves the cue point issue for you too.

JoergAtGithub committed:
Handled the cue point case. Don't show beat indication after (hot) cue button press.

JoergAtGithub committed:
Changed reverse beat indication state from -1 to +2

JoergAtGithub committed:
Merge branch 'main' into improve_beat_indicator

JoergAtGithub committed:
pre-commit

JoergAtGithub committed:
Improved responsiveness when pressing play/cue button at while track is hold on a beat

JoergAtGithub commented:
@daschuer Did you tried the latest version of this PR? From my point of view, everything is working now.

daschuer reviewed (approved):
Works perfectly now.

Thank you very much.

JoergAtGithub commented:
Merge?

daschuer commented:
Ups, must have missed to hit merge.
